### PR TITLE
Save runtimes in configmaps

### PR DESCRIPTION
**Pull request by rakesh-garimella**
Work in Progress

### CODE REVIEW

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -88,6 +89,9 @@ func New(cfg Config, smclient *monitoringv1alpha1.MonitoringV1alpha1Client) *Con
 		},
 	})
 
+	var l langruntime.Langruntimes
+	l.ReadConfigMap(cfg.KubeCli)
```
Comment (andresmgot) [inline pkg/controller/controller.go]: `ReadConfigMap` is too generic when it is retrieving the runtime info. Also the function should retrieve more info like the namespace of the configmap and the name. My suggestion is to rename it as the method `New` of the `Langruntimes`:

```
var l = langruntimes.New(cfg.KubeCli, 'kubeless', 'kubeless-config')
```

File: pkg/langruntime/langruntime.go [hunk 2]
```diff
@@ -7,83 +7,71 @@ import (
...
-	versions       []runtimeVersion
-	DepName        string
-	FileNameSuffix string
-}
+const (
+	pubsubFunc = "PubSub"
+)
 
-var pythonVersions, nodeVersions, rubyVersions, dotnetcoreVersions []runtimeVersion
 var availableRuntimes []RuntimeInfo
```
Comment (andresmgot) [inline pkg/langruntime/langruntime.go]: `availableRuntimes` should be a property of the Langruntimes struct

File: pkg/langruntime/langruntime.go [hunk 3]
```diff
@@ -7,83 +7,71 @@ import (
...
+	Version          string         `yaml:"version"`
+	HTTPImage        string         `yaml:"httpImage"`
+	PubSubImage      string         `yaml:"pubsubImage"`
+	InitImage        string         `yaml:"initImage"`
+	ImagePullSecrets []ImageSecrets `yaml:"imagePullSecrets,omitempty"`
+}
+
+// ImageSecrets for pulling the image
+type ImageSecrets struct {
+	ImageSecret string `yaml:"imageSecret,omitempty"`
```
Comment (andresmgot) [inline pkg/langruntime/langruntime.go:43]: better call it 
```
ImageSecret string `yaml:"imageSecret,omitempty"` 
```
so it has a clear name for its purpose

File: pkg/langruntime/langruntime.go [hunk 4]
```diff
@@ -7,83 +7,71 @@ import (
...
+// and the supported versions
+type RuntimeInfo struct {
+	ID             string           `yaml:"ID"`
+	Versions       []runtimeVersion `yaml:"versions"`
+	DepName        string           `yaml:"depName"`
+	FileNameSuffix string           `yaml:"fileNameSuffix"`
+}
+
 // GetRuntimes returns the list of available runtimes as strings
 func GetRuntimes() []string {
```
Comment (andresmgot) [inline pkg/langruntime/langruntime.go]: all the methods should now be part of the `Langruntime` struct:
```
func (l *Langruntime) GetRuntimes() []string{
 	result := []string{}
  	for _, runtimeInf := range l.availableRuntimes {
...
````

File: pkg/langruntime/langruntime.go [hunk 5]
```diff
@@ -7,83 +7,71 @@ import (
...
-		{ID: "nodejs", versions: nodeVersions, DepName: "package.json", FileNameSuffix: ".js"},
-		{ID: "ruby", versions: rubyVersions, DepName: "Gemfile", FileNameSuffix: ".rb"},
-		{ID: "dotnetcore", versions: dotnetcoreVersions, DepName: "requirements.xml", FileNameSuffix: ".cs"},
+	if err != nil {
+		logrus.Fatalf("Unable to get the runtime images. %v", err)
 	}
 }
 
+type runtimeVersion struct {
+	Name             string         `yaml:"name"`
```
Comment (andresmgot) [inline pkg/langruntime/langruntime.go]: try to avoid to export properties if they are not going to be accessible from the outside

File: pkg/langruntime/langruntime.go [hunk 6]
```diff
@@ -7,83 +7,71 @@ import (
...
+	Version          string         `yaml:"version"`
+	HTTPImage        string         `yaml:"httpImage"`
+	PubSubImage      string         `yaml:"pubsubImage"`
+	InitImage        string         `yaml:"initImage"`
+	ImagePullSecrets []ImageSecrets `yaml:"imagePullSecrets,omitempty"`
+}
+
+// ImageSecrets for pulling the image
+type ImageSecrets struct {
```
Comment (andresmgot) [inline pkg/langruntime/langruntime.go:42]: the same for the types, do not export them if they are not going to be used in a different package

File: pkg/utils/k8sutil.go [hunk 7]
```diff
@@ -684,6 +684,19 @@ func EnsureFuncDeployment(client kubernetes.Interface, funcObj *kubelessApi.Func
 		}
 		dpm.Spec.Template.Spec.InitContainers = []v1.Container{provisionContainer}
 	}
+
+	// Add the imagesecrets if present to pull images from private docker registry
+	imageSecrets, err := langruntime.GetImageSecrets(funcObj.Spec.Runtime)
+
```
Comment (andresmgot) [inline pkg/utils/k8sutil.go]: you missed the `if err {...` here

File: pkg/utils/k8sutil.go [hunk 8]
```diff
@@ -684,6 +684,19 @@ func EnsureFuncDeployment(client kubernetes.Interface, funcObj *kubelessApi.Func
...
+	if len(imageSecrets) > 0 {
+		var lors []v1.LocalObjectReference
+		for _, s := range imageSecrets {
+			lor := v1.LocalObjectReference{Name: s}
```
Comment (andresmgot) [inline pkg/utils/k8sutil.go]: It is better for `GetImageSecrets` to return a `[]v1.LocalObjectReference` rather than an array of strings.

**Review by andresmgot — COMMENTED** [review]
This is going the right way. Thank you for working on this!

See my inline comments and let me know if you have any question.

Also note that in the CLI we are using the `langruntime` package as well (cmd/kubeless/function/deploy and update). In that case I think we should try to obtain the ConfigMap but since it is used just to list the available runtime when running `--help` I would not throw an error (the `kubeless` namespace may be protected and be inaccessible from the client side).

File: cmd/kubeless/function/update.go [hunk 9]
```diff
@@ -170,7 +174,10 @@ var updateCmd = &cobra.Command{
 }
 
 func init() {
-	updateCmd.Flags().StringP("runtime", "", "", "Specify runtime. Available runtimes are: "+strings.Join(langruntime.GetRuntimes(), ", "))
+	cli := utils.GetClientOutOfCluster()
+	var lr = langruntime.New(cli, "kubeless", "kubeless-config")
```
Comment (sayanh) [inline cmd/kubeless/function/update.go:178]: General comment: So we have to restart controller to change a runtimes configmap as it is picked up during bootup. Is this the expected behavior?

File: cmd/kubeless/function/update.go [hunk 10]
```diff
@@ -61,9 +66,9 @@ var updateCmd = &cobra.Command{
 			logrus.Fatal(err)
 		}
 
-		if runtime != "" && !langruntime.IsValidRuntime(runtime) {
+		if runtime != "" && !lr.IsValidRuntime(runtime) {
```
Comment (sayanh) [inline cmd/kubeless/function/update.go]: Personally, I would love `langruntime` over `lr`

File: cmd/kubeless/function/function.go [hunk 11]
```diff
@@ -258,7 +258,7 @@ func getFunctionDescription(cli kubernetes.Interface, funcName, ns, handler, fil
 	svcSpec := v1.ServiceSpec{
 		Ports: []v1.ServicePort{
 			{
-				Name:     "function-port",
+				Name:     "http-function-port",
```
Comment (sayanh) [inline cmd/kubeless/function/function.go]: Looks like git manipulation is needed as `http-` is not part of this change

File: examples/Makefile [hunk 12]
```diff
@@ -239,6 +239,7 @@ pubsub-python-verify:
 	$$found
 
 pubsub-python34:
+	kubeless topic create s3-python34 || true
```
Comment (sayanh) [inline examples/Makefile]: Why is this needed?
